File: benchmarks/agents/memory/DataBuffer.py
from benchmarks import benchmarks

import collections
import benchmarks
import math
import torch

from benchmarks.agents.memory.PriorityTree import PriorityTree


class DataBuffer:
    """
    A buffer allowing for storage and retrieval of experience datum (i.e., action, reward, done, and priority).
    """

    def __init__(self, capacity, n_steps, gamma, initial_priority, n_children):
        """
        Create a data buffer.
        :param capacity: the number of experiences the buffer can store
        :param n_steps: the number of steps for which rewards are accumulated in multistep Q-learning
        :param gamma: the discount factor
        :param initial_priority: the initial priority given to first elements
        :param n_children: the number of children each node has
        """

        # Store the data buffer's parameters.
        self.capacity = capacity
        self.n_steps = n_steps
        self.gamma = gamma

        # Queues keeping track of past actions, cumulated rewards, and dones.
        self.past_actions = collections.deque(maxlen=n_steps)
        self.past_rewards = collections.deque(maxlen=n_steps)
        self.past_dones = collections.deque(maxlen=n_steps)

        # Torch tensors storing all the buffer's data.
        self.device = benchmarks.device()
        self.actions = torch.zeros([capacity], dtype=torch.int).to(self.device)
        self.rewards = torch.zeros([capacity], dtype=torch.float).to(self.device)
        self.dones = torch.zeros([capacity], dtype=torch.bool).to(self.device)

        # The priorities associated with all experiences in the replay buffer.
        self.priorities = PriorityTree(capacity=capacity, initial_priority=initial_priority, n_children=n_children)

        # The index of the next datum to add in the buffer.
        self.current_id = 0

    def append(self, experience):
        """
        Add the datum of the next experience to the buffer.
        :param experience: the experience whose datum must be added to the buffer
        """

        # Update the returns of last experiences.
        for i in range(len(self.past_rewards)):
            self.past_rewards[i] += math.pow(self.gamma, i + 1) * experience.reward

        # Add the reward, action and done to their respective queues.
        self.past_rewards.appendleft(experience.reward)
        self.past_actions.appendleft(experience.action)
        self.past_dones.appendleft(experience.done)

        # Add new data to the buffer.
        if experience.done is True:

            # If the current episode has ended, keep track of all valid data.
            while len(self.past_rewards) != 0:
                self.add_datum(self.past_actions.pop(), self.past_rewards.pop(), self.past_dones[0])

            # Then, clear the queues of past reward, actions, and dones.
            self.past_rewards.clear()
            self.past_actions.clear()
            self.past_dones.clear()

        elif len(self.past_rewards) == self.n_steps:

            # If the current episode has not ended, but the queues are full, then keep track of next valid datum.
            self.add_datum(self.past_actions.pop(), self.past_rewards.pop(), self.past_dones[0])

    def __getitem__(self, indices):
        """
        Retrieve the data of the experience whose indices are passed as parameters.
        :param indices: the indices of the experiences whose data must be retrieved
        :return: the data (i.e., action at time t, n-steps return at time t, and done at time t + n_steps)
        """
        if self.current_id >= self.capacity:
            indices = [(index + self.current_id) % self.capacity for index in indices]
        return self.actions[indices], self.rewards[indices], self.dones[indices]

    def __len__(self):
        """
        Retrieve the number of experiences stored in the buffer.
        :return: the number of experiences stored in the buffer
        """
        return min(self.current_id, self.capacity)

    def clear(self):
        """
        Empty the data buffer.
        """
        self.past_actions.clear()
        self.past_rewards.clear()
        self.past_dones.clear()
        self.actions = torch.zeros([self.capacity])
        self.rewards = torch.zeros([self.capacity])
        self.dones = torch.zeros([self.capacity])
        self.priorities.clear()
        self.current_id = 0

    def add_datum(self, action, reward, done):
        """
        Add a datum to the buffer.
        :param action: the action at time t
        :param reward: the n-steps reward
        :param done: the done at time t + n_steps
        """
        index = self.current_id % self.capacity
        self.actions[index] = action
        self.rewards[index] = reward
        self.dones[index] = done
        self.priorities.append(self.priorities.max())
        self.current_id += 1

File: benchmarks_cpp/src/data_buffer.cpp
#include <cmath>
#include "replay_buffer.hpp"
#include "data_buffer.hpp"

DataBuffer::DataBuffer(int capacity, int n_steps, float gamma, float initial_priority, int n_children)
    : past_actions(n_steps), past_rewards(n_steps), past_dones(n_steps), device(ReplayBuffer::getDevice()) {

    // Store the data buffer's parameters.
    this->capacity = capacity;
    this->n_steps = n_steps;
    this->gamma = gamma;

    // Torch tensors storing all the buffer's data.
    this->actions = torch::zeros({capacity}, at::kInt).to(this->device);
    this->rewards = torch::zeros({capacity}, at::kFloat).to(this->device);
    this->dones = torch::zeros({capacity}, at::kBool).to(this->device);

    // The priorities associated with all experiences in the replay buffer.
    this->priorities = std::make_unique<PriorityTree>(capacity, initial_priority, n_children);

    // The index of the next datum to add in the buffer.
    this->current_id = 0;
}

void DataBuffer::append(ExperienceTuple experience_tuple) {

    Experience experience = Experience(experience_tuple);

    // Update the returns of last experiences.
    for (std::uint64_t i = 0; i < this->past_rewards.size(); i++) {
        this->past_rewards[i] += std::pow(this->gamma, i + 1) * experience.reward;
    }

    // Add the reward, action and done to their respective queues.
    this->past_rewards.push_front(experience.reward);
    this->past_actions.push_front(experience.action);
    this->past_dones.push_front(experience.done);

    // Add new data to the buffer.
    if (experience.done == true) {

        // If the current episode has ended, keep track of all valid data.
        while (this->past_rewards.size() != 0) {
            this->addDatum(this->past_actions.back(), this->past_rewards.back(), this->past_dones[0]);
            this->past_actions.pop_back();
            this->past_rewards.pop_back();
        }

        // Then, clear the queues of past reward, actions, and dones.
        this->past_rewards.clear();
        this->past_actions.clear();
        this->past_dones.clear();

    } else if (static_cast<int>(this->past_rewards.size()) == this->n_steps) {

        // If the current episode has not ended, but the queues are full, then keep track of next valid datum.
        this->addDatum(this->past_actions.back(), this->past_rewards.back(), this->past_dones[0]);
    }
}

std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> DataBuffer::operator[](torch::Tensor indices) {
    if (this->current_id >= this->capacity) {
        indices = torch::remainder(indices + this->current_id, this->capacity);
    }
    return std::make_tuple(this->actions.index({indices}), this->rewards.index({indices}), this->dones.index({indices}));
}

int DataBuffer::size() {
    return std::min(this->current_id, this->capacity);
}

void DataBuffer::clear() {
    this->past_actions.clear();
    this->past_rewards.clear();
    this->past_dones.clear();
    this->actions = torch::zeros({capacity}, at::kInt).to(this->device);
    this->rewards = torch::zeros({capacity}, at::kFloat).to(this->device);
    this->dones = torch::zeros({capacity}, at::kBool).to(this->device);
    this->priorities->clear();
    this->current_id = 0;
}

void DataBuffer::addDatum(int action, float reward, bool done) {
    int index = this->current_id % this->capacity;
    this->actions.index_put_({index}, action);
    this->rewards.index_put_({index}, reward);
    this->dones.index_put_({index}, done);
    this->priorities->append(this->priorities->max());
    this->current_id += 1;
}

std::unique_ptr<PriorityTree> &DataBuffer::getPriorities() {
    return this->priorities;
}
